### DISCUSSION TITLE
Shell integration - ssh fallback to xterm-256color

### DISCUSSION THREAD

**Discussion by tekumara** [ghostty-org/ghostty#2701] [Feature Requests, Ideas] [answered]
When:
1. ssh hosts can't be configured with the `xterm-ghostty` terminfo
2. it's not possible or desirable to configure the [.ssh/config fallback](https://github.com/ghostty-org/ghostty?tab=readme-ov-file#configure-ssh-to-fall-back-to-a-known-terminfo-entry) eg: because it overrides other terminal emulators (kitty) 

It would be nice to be able to enable a shell integration that aliased `ssh` to `TERM=xterm-256color ssh`.  

It could also help ghostty work with [zero config out of the box](https://github.com/ghostty-org/ghostty/discussions/2363) (although there is a tradeoff here with making it work optimally, by installing the `xterm-ghostty` terminfo)

**Comment by mitchellh** [comment, marked as answer]
This is a chicken and egg problem due to Ghostty being new. This will be resolved after public release: https://github.com/ghostty-org/ghostty/issues/2542

We don’t want to say we’re xterm-256color because we’re not. This is a bandaid that happens to work but it’s a lie, causes a number of Ghostty features to become unavailable, and perpetuates TERM being used incorrectly. 

The proper solution is #2542 but even better for programs to use XTGETTCAP instead which makes terminfo work over SSH. I talk about this in my devlogs a bit.

Slightly related, I think there’s opportunity for something like the Kitty ssh kitten but this solves a different problem and I still wouldn’t want to override TERM: https://sw.kovidgoyal.net/kitty/kittens/ssh/

**Reply by lucas-yotsui** [reply]
Something like **ssh kitten** would be really nice!! 
I've used Kitty up until Ghostty's public release and that's the only thing I'm missing so far

**Reply by 3v1n0** [reply]
I like the approach of creating this "issue" by design, since indeed everyone just accepted to rely on `xterm` fallback, making all the other apps not to adapt and eventually getting a worse terminal experience for no reasons.

However, indeed we should probably have some smarter way to figure out whether the application we're going to use is ready for this or not, and fallback when necessary (even with a warn).

For example, inside snaps (and likely flatpak) due to confinement, no application you can launch with ghostty is able, by default, to read the content of ghostty TERMCAP, thus it won't work... Instead ghostty (or maybe the snap launcher?) should use the desktop portal APIs to export the file in a place the application can read and set the env variable accordingly.

So, while applications should now about `xterm-ghostty`, there will always be a legacy app out there that won't ever be able to use it.

**Comment by 3v1n0** [comment]
A simple workaround for affected apps is having something like this in your shell profile though (at least in zsh and bash this works fine):

```bash
if [ "$TERM" = xterm-ghostty ]; then
  fallback_programs=(
    ssh
    matterhorn
    snap
    flatpak
  )

  for cmd in "${fallback_programs[@]}"; do
    real_cmd=$(command -v "$cmd" 2>/dev/null)
    if [ -n "$cmd" ]; then
      alias $cmd="env TERM=xterm-256color $real_cmd"
    fi
  done
fi
```

**Reply by rhodes-b** [reply]
Ssh isnt required 1.2 supports a shell integration option 

`shell-integration-features = ssh-env`

Any other commands are at fault of themselves because ghostty terminfo should otherwise be available on your system so they shouldnt be using some bundled terminfo
